### PR TITLE
Updated mp_builtin_pow() to use the optimised mpz_pow3_inpl() function.

### PR DESCRIPTION
**Pull request by nickovs**
I took a stab at fixing #2774 so that the built in pow() function is usable with arguments large enough to be useful for cryptographic functions, and to not use exponential memory and time. All it does is make use of the existing (but previously commented out) `mpz_pow3_inpl()` function written by @dpgeorge.

This is my first pull request for Micropython to please feel free to send me any comments and complaints and I will try to fix them!

### CODE REVIEW

**Comment by nickovs** [comment]
It seems that the Travis CI build is currently failing with:
```
$ (cd tests && for exp in *.exp; do testbase=$(basename $exp .exp); echo -e "\nFAILURE $testbase"; diff -u $testbase.exp $testbase.out; done)

FAILURE *
diff: *.exp: No such file or directory
diff: *.out: No such file or directory
```
I didn't touch anything in the tests since the existing tests covered the new code. I see that @dpgeorge's #2795 is failing in the same way, so I'm pretty sure it's not my fault! That said, I don't know anything about the Travis setup here so I don't know what I can do to fix this.

**Comment by pfalcon** [comment]
Finding a cause of a Travis failure is a bit of black art, as it doesn't have quick-links to potential errors (or we don't have it configured). How I usually do it is grepping its log for "error" (many false positives) or "exited with" (not selective enough, scroll back until you find the issue).

**Comment by nickovs** [comment]
OK, I found the error. The Travis build is clean now.

**Comment by dpgeorge** [comment]
Thanks for the contribution!

A few general comments:
- please use only spaces (no tabs) and use a "tab width" of 4
- if statements should (almost) always use curly braces, even for 1-liners
- variable declarations should be as close as possible to their first use
- it would be great if there were tests for this, even better if it had 100% coverage (can be checked via the coveralls link in the CI output above

**Comment by nickovs** [comment]
OK. I have adjusted the code style to have 4 spaces, no tabs and braces on single-statement blocks and I also moved the variable definitions to be as close to use as made sense.

I also added comprehensive test coverage including the exception cases; coveralls now reports 100% on modbuiltins.c. Let me know if there is anything else you would like me to change.

File: py/modbuiltins.c [hunk 1]
```diff
@@ -375,10 +375,44 @@ STATIC mp_obj_t mp_builtin_ord(mp_obj_t o_in) {
...
 STATIC mp_obj_t mp_builtin_pow(size_t n_args, const mp_obj_t *args) {
     switch (n_args) {
         case 2: return mp_binary_op(MP_BINARY_OP_POWER, args[0], args[1]);
-        default: return mp_binary_op(MP_BINARY_OP_MODULO, mp_binary_op(MP_BINARY_OP_POWER, args[0], args[1]), args[2]); // TODO optimise...
+#if MICROPY_LONGINT_IMPL != MICROPY_LONGINT_IMPL_MPZ
+        default: return mp_binary_op(MP_BINARY_OP_MODULO, mp_binary_op(MP_BINARY_OP_POWER, args[0], args[1]), args[2]);
+#else
+        default:
+	    if (!MP_OBJ_IS_INT(args[0]) || !MP_OBJ_IS_INT(args[1]) || !MP_OBJ_IS_INT(args[2])) {
+	        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "pow() with 3 arguments requires integers"));
```
Comment (dpgeorge) [inline py/modbuiltins.c]: A shortcut function to use is mp_raise_TypeErro("...msg...").

File: py/modbuiltins.c [hunk 2]
```diff
@@ -375,10 +375,44 @@ STATIC mp_obj_t mp_builtin_ord(mp_obj_t o_in) {
...
+	    } else {
+                mp_obj_t result = mp_obj_new_int_from_ull(0); // Use the _from_ull version as this forces an mpz int
+                mp_obj_int_t *res_p = (mp_obj_int_t *) MP_OBJ_TO_PTR(result);
+
+	        mpz_t l_temp, r_temp, m_temp;
+	        mpz_t *lhs = mp_mpz_for_int(args[0], &l_temp);
```
Comment (dpgeorge) [inline py/modbuiltins.c]: There's something a bit wrong with indenting here...

File: tests/basics/builtin_pow.py [hunk 3]
```diff
@@ -8,4 +8,35 @@
...
+    # This value happens to be prime
+    x = 0xd48a1e2a099b1395895527112937a391d02d4a208bce5d74b281cf35a57362502726f79a632f063a83c0eba66196712d963aa7279ab8a504110a668c0fc38a7983c51e6ee7a85cae87097686ccdc359ee4bbf2c583bce524e3f7836bded1c771a4efcb25c09460a862fc98e18f7303df46aaeb34da46b0c4d61d5cd78350f3edb60e6bc4befa712a849
+    y = 0x3accf60bb1a5365e4250d1588eb0fe6cd81ad495e9063f90880229f2a625e98c59387238670936afb2cafc5b79448e4414d6cd5e9901aa845aa122db58ddd7b9f2b17414600a18c47494ed1f3d49d005a5
+    print(hex(pow(2, 200, x))) # Should not overflow, just 1 << 200
+    print(hex(pow(2, x-1, x))) # Should be 1, since x is prime
+    print(hex(pow(y, x-1, x))) # Should be 1, since x is prime
+    print(hex(pow(y, y-1, x))) # Should be a 'big value'
+    print(hex(pow(y, y-1, y))) # Should be a 'big value'
+except OverflowError:
+    print("OverflowError, expected for builds without mpz support")
```
Comment (dpgeorge) [inline tests/basics/builtin_pow.py]: The test suite currently doesn't handle the case when mpz is disabled.  So you can just remove the try-except block here.

(If we do ever support tests without mpz, then such tests need to be factored into separate files, so the whole file is skipped.)

**Review by dpgeorge — COMMENTED** [review]
Can you please also add a config option, eg `MICROPY_OPT_MPZ_POW3`, to enable/disable this?  It would be disabled by default.

**Comment by pfalcon** [comment]
> Can you please also add a config option, eg MICROPY_OPT_MPZ_POW3, to enable/disable this? It would be disabled by default.

I thought about an option to disable pow() altogether, as 2-arg variant can be replaced by `**` operator.

**Comment by dpgeorge** [comment]
> I thought about an option to disable pow() altogether,

pow is pretty basic and used a lot... such an option would only be for really constrained ports.

File: py/modbuiltins.c (hunk 1 above)
Comment (nickovs) [inline py/modbuiltins.c]: Done

File: py/modbuiltins.c (hunk 2 above)
Comment (nickovs) [inline py/modbuiltins.c]: Fixed.

File: tests/basics/builtin_pow.py (hunk 3 above)
Comment (nickovs) [inline tests/basics/builtin_pow.py]: Done.

**Comment by nickovs** [comment]
> Can you please also add a config option, eg MICROPY_OPT_MPZ_POW3, to enable/disable this? It would be disabled by default.

I can certainly do that. That said, the current original implementation throws an out-of-memory exception for anything other than trivially small exponents if you have MICROPY_LONGINT_IMPL_MPZ enabled. It is safe with MICROPY_LONGINT_IMPL_NONE and MICROPY_LONGINT_IMPL_LONGLONG since they silently truncate the intermediate results. Furthermore, you elsewhere wrote:

> The test suite currently doesn't handle the case when mpz is disabled. So you can just remove the try-except block here.

Since I've added proper test coverage for the builtins.pow() the old implementation will fail the tests since we are now testing `pow(x,y,z)` with meaningful-sized arguments. This means that if the proposed configuration option is in its default state the tests will fail!

**Comment by nickovs** [comment]
For what it's worth, disabling the new code on the stmhal build takes the size of firmware.elf from 636124 to 636062, a saving of 62 bytes or a little less than 0.01%.

I'm happy to put in a configuration option to disable this code but given that doing this will make reasonable inputs throw MemoryError exceptions I'd be inclined to have the switch (a) disable 3-argument pow() altogether and (b) be off by default. Having 3-argument pow() and multi-precision integers both enabled without having the efficient implementation switched on doesn't seem like a good plan to me.

**Comment by pfalcon** [comment]
> This means that if the proposed configuration option is in its default state the tests will fail!

@dpgeorge meant that its default state in py/mpconfig.h should be false. We can enable it for Unix port if you insist ;-).

> For what it's worth, disabling the new code on the stmhal build takes the size of firmware.elf from 636124 to 636062, a saving of 62 bytes or a little less than 0.01%.

ELF files contain a lot of alignment, etc. Valid comparison would be on raw .bin files (or on size(1) called on .elf).

**Comment by nickovs** [comment]
OK. The difference in sizes for the firmware1.bin and firmware.dfu files is 392 bytes, which is just over 0.1%.

I am still of the opinion that if you are going to include a whole multi-precision integer package in the build then it should not crash with MemoryError exceptions when fed reasonable inputs. The old 3-arg pow() implementation works just fine when you compile without mpz support; the code here only enables the new version when you have mpz. I'd be fine with a config option that disables 3-arguement pow() altogether since the result is reliable. It just seems wrong to me to have a configuration for "Enable an unreliable implementation that may or may no crash depending on your inputs", which is what the old implementation does when mpz support is present.

**Comment by stinos** [comment]
@nickovs can you also check https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md and make your commits adhere to it? E.g. something like 'py: Update xyz' instead of 'Updated xyz' And make the body of the commit message explain *why* the change is needed. You say 'Added MICROPY_MPZ_POW3 configuration option etc' but without stating why anyone would need that. The reason is in the comments above but should really be explained as part of the commit as that's much handier for future reference. Also take the second commit for example: the body says 'Removed try/except around MPZ pow() tests'. That is clear by reading the diff so it's merely duplicate information. Instead it should say why it's somehow ok to just remove the exception handling. Also you should rebase/squash some commits so they create an easy to follow 'story'. Can till be seperate commits, but it should mean something. 'Fix typo' commits should definitely be squashed.

**Comment by pfalcon** [comment]
@nickovs : Looking in more detail, there're more issues with this patch. The biggest one is that it leaks implementation detail of a particular object implementation into the generic module, modbuiltins.c. That module doesn't know anything about mpz_t, and shouldn't know, and even wrapped in in #if's, it's not pretty. Sorry about this. The good news is that we appreciate contributions to this project, and ready to support first-time contributors by cooperating on patch implementation, by picking off on the hard parts. 

Actually, I was ready to do refactoring myself now, but decided to ask if you have time and interest to go thru the whole process yourself actually. Note that you didn't pick up the easiest task to start with, even despite it may seem so. That's because of the high configurability of MicroPython, and specifically of this part (integer object implementation). I hit the same issue with .from_bytes() implementation - even with my experience in uPy, it took more chore than I initially expected. So, I'd recommend to handover it now to a maintainer.

**Comment by nickovs** [comment]
@pfalcon Thank you for the feedback. I did think about that when I started the code but it seemed like if it was conditional on the configuration than this was as good as any place to put it and it would save a layer of function calls. I will move the code to objint_mpz.c, since that clearly has knowledge of mpz_t.

**Comment by pfalcon** [comment]
That's the right place for it, thanks. While you refactor it, also consider applying suggestions above on the commit squashing and format of commit messages (you're welcome to add individual commits to PR if it helps to track your work while it's WIP, but for the final merge, they need to be squashed; if a contributor does that, it saves few minutes to maintainers (which usually amounts to be a patch being merged sooner when it's ready)).

**Comment by pfalcon** [comment]
@nickovs : FYI, we use rebase-based workflow (just like (most of maintainers of) Linux and many other projects nowadays). Just thought I'd mention, because it's actually easier for contributors too.

**Comment by nickovs** [comment]
OK, I've moved the mpz-specifc code into objint_mpz.c and put in place a configuration switch to disable 3-argument pow() for people who need to save the space. I also (think I have) squashed the commits so that it should be a single big commit with a meaningful, style-conformant commit message. Hopefully this will tick all the boxes!

**Comment by pfalcon** [comment]
> I also (think I have) squashed

Something went wrong:

>  Commits 55   Files changed 100

Merge commits are to blame, they're the root of all evil. Never "git pull", always "git pull --rebase" ;-)

**Comment by nickovs** [comment]
OK. What should I do now?

**Comment by pfalcon** [comment]
If you actually squashed your changes into 1 commit, it should be among those 55 I guess. I'd look for it, created a fresh branch from master, cherry-picked that commit and re-pushed. If squash didn't work out, there would be multiple commits to cherry-pick and then squash. As usual, I'm happy to offer assistance with this git grunge.

**Comment by nickovs** [comment]
OK, rather than try to wrestle with git, I have created a new branch and new pull request #2835 which should be clean. I suggest we close this one and move the discussion to there.

**Comment by pfalcon** [comment]
You have the powers, please do.

**Pull request closed without merging**